### DISCUSSION TITLE
Selectively disable fused-multiply add (FMA) instructions on GPU?

### DISCUSSION THREAD

**Discussion by BenWibking** [AMReX-Codes/amrex#2120] [Q&A] [answered]
Hi,

I am benchmarking my hydrodynamics code that uses AMReX. I have noticed that the default CUDA floating point compiler settings lead to a directional asymmetry in 2D problems that should be exactly symmetric. Setting `--fmad=false` fixes the issue and leads to results that are exactly symmetric to every digit in 2D.

However, disabling FMA causes a ~30% performance hit. It appears that I only need to disable FMA instructions for specific kernels that combine information from multiple directions, while leaving the Riemann solver kernels free to use FMA. Is this possible to do in CUDA (or any other architecture)?

**Comment by WeiqunZhang** [comment, marked as answer]
In GNU Make, the rule for compiling `.cpp` files is here https://github.com/AMReX-Codes/amrex/blob/development/Tools/GNUMake/Make.rules#L197.  It's a generic rule.  You can add your own rule in your own make file to override it for a specific file.  For example,
```
$(objEXETempDir)/riemann.o: riemann.cpp
    $(CXX) ...... 
```
@ax3l can probably tell you the proper way of doing this in cmake.

**Reply by BenWibking** [reply]
I am currently only setup to build with CMake for this code, so if @ax3l could point me to the documentation on how to do this in CMake, that would be very helpful.

**Reply by BenWibking** [reply]
Ah, nevermind, this actually won't work for my code. All of my kernels are defined by function templates instantiated within (as it turns out) a single .cpp file, so file-by-file setting of the compilation options doesn't help my code.

**Reply by ax3l** [reply]
Ah I see, then pragmas (below) are probably the right way to be selective.

For the record, _individual file_ flags can be defined via `set_source_files_properties` in CMake, e.g.
```cmake
set_source_files_properties(foo.cpp PROPERTIES COMPILE_FLAGS -Wno-effc++)
```
https://stackoverflow.com/a/13639476/2719194

The flag value can be potentially combined with a generator expression, filtering only for specific compilers that understand the flag.

**Comment by WeiqunZhang** [comment]
GCC does have a pragma for controlling optimization of individual functions. https://gcc.gnu.org/onlinedocs/gcc/Function-Specific-Option-Pragmas.html#Function-Specific-Option-Pragmas.  But nvcc may not have it.  @maxpkatz

**Reply by BenWibking** [reply]
If there is an nvcc `__attribute__` for this as well, could you consider creating an AMReX-portable version by adding a AMREX_NO_FMA macro?

**Reply by BenWibking** [reply]
For gcc on x86-64, it appears that the easiest way to do this is to add an attribute to the function declaration:
`__attribute__ ((__target__ ("no-fma")))
`

**Reply by BenWibking** [reply]
Just for reference, this attribute is also supported by Clang when building for x86.

**Reply by maximumcats** [reply]
As far as I am aware, nvcc does not have equivalent functionality.

**Reply by BenWibking** [reply]
Ah, that's unfortunate :/ Is there a way to submit a feature request for this? It would be extremely useful for me, but probably also for many AMReX codes.

**Comment by philip-blakely** [comment]
From the NVCC documentation, you could write the multiply/add via intrinsics:
https://docs.nvidia.com/cuda/floating-point/index.html#controlling-fused-multiply-add (Section 4.3)
although this seems time consuming to put in place, and will make your code less readable.

**Reply by BenWibking** [reply]
Yes, I agree that this is not a very practical solution.
